mt_gateway_test.go

```diff
@@ -0,0 +1,382 @@
+package main
+
+import (
+	"context"
+	b64 "encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/ory/dockertest/v3"
+	"github.com/phayes/freeport"
+	"github.com/stretchr/testify/require"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/rudderlabs/rudder-server/app"
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/testhelper/destination"
+	thEtcd "github.com/rudderlabs/rudder-server/testhelper/etcd"
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+	"github.com/rudderlabs/rudder-server/testhelper/rand"
+	whUtil "github.com/rudderlabs/rudder-server/testhelper/webhook"
+	"github.com/rudderlabs/rudder-server/utils/types/deployment"
+)
+
+func TestMultiTenantGateway(t *testing.T) {
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() { <-c; cancel() }()
+	defer cancel()
+
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	var (
+		group             errgroup.Group
+		etcdContainer     *thEtcd.Resource
+		postgresContainer *destination.PostgresResource
+		serverInstanceID  = "1"
+	)
+
+	group.Go(func() (err error) {
+		postgresContainer, err = destination.SetupPostgres(pool, t)
+		return err
+	})
+	group.Go(func() (err error) {
+		etcdContainer, err = thEtcd.Setup(pool, t)
+		return err
+	})
+	require.NoError(t, group.Wait())
+
+	webhook := whUtil.NewRecorder()
+	t.Cleanup(webhook.Close)
+
+	writeKey := rand.String(27)
+	workspaceID := rand.String(27)
+	workspaces := map[string]backendconfig.ConfigT{
```

Comment:
I would prefer to see workspace config in a json file, inside testdata folder.

Reply:
@lvrach there you [go](https://github.com/rudderlabs/rudder-server/pull/2078/commits/4e7754bb432716adf307b1451e09997fdbc2de69)